### ISSUE TITLE
missing `requires_dist` in PyPI JSON makes installation slow

### ISSUE THREAD

**Issue by aryarm** [statsmodels/statsmodels#8200]
## Bug description
Python package managers like `poetry` (and probably `pip` now that it resolves dependencies [as of v20.3](https://pyfound.blogspot.com/2020/11/pip-20-3-new-resolver.html)) depend on packages listing their dependencies in the `requires_dist` section of the JSON returned by the PyPI API. Most packages provide this information, but the `statsmodels` package does not.

```
> curl -s https://pypi.org/pypi/scipy/json | jq ".info.requires_dist"
[
  "numpy (<1.25.0,>=1.17.3)"
]
> curl -s https://pypi.org/pypi/patsy/json | jq ".info.requires_dist"
[
  "six",
  "numpy (>=1.4)",
  "pytest ; extra == 'test'",
  "pytest-cov ; extra == 'test'",
  "scipy ; extra == 'test'"
]
> curl -s https://pypi.org/pypi/statsmodels/json | jq ".info.requires_dist"
null
```

Without this information, package mangers must download and inspect the `whl`s of every version of `statsmodels` in order to extract their dependencies and then resolve them against any current constraints. This can make installation of `statsmodels` very slow; it took me ~2.5 hours to install it using `poetry`.

## References
See https://github.com/aws/aws-cli/issues/5701 for a similar situation in which this was a problem. See [the `poetry` documentation](https://python-poetry.org/docs/faq/#why-is-the-dependency-resolution-process-slow) and https://github.com/python-poetry/poetry/issues/2094 for a more detailed explanation of why this is a problem for `poetry`. See https://github.com/pypa/pip/issues/9187#issuecomment-736318672 for a discussion of why this is a problem for `pip`.

## Cause of bug
Based on steps 6 and 8 of [the `statsmodels` team maintainer notes](https://github.com/statsmodels/statsmodels/blob/408eae829c8b3ebca2f0ea46e69f3a62b94a554d/docs/source/dev/maintainer_notes.rst), it seems like the dev team uses `twine` to upload an sdist to PyPI before it uploads its wheels, so https://github.com/pypa/twine/issues/761 may be the cause of the issue.

The solution in that case is pretty simple: just change the order. Upload the wheels before the sdists, instead.

**Comment by bashtage** [comment]
Historically this has been done because of worries about statsmodels development upgrading other binaries, e.g., SciPy.  I suspect that the changes over the past years have mostly mitigated these worries. 

@josef-pkt what do you think about including requirements in setup?

**Comment by josef-pkt** [comment]
Is the default of `pip install statsmodels` still to automatically update dependencies?

I always hated it when pip tried to upgrade numpy and scipy and destroyed my dev setting.

**Comment by bashtage** [comment]
It doesn't if they satisfy the requirements. It updates to the latest supported if they do usually.

**Comment by aryarm** [comment]
> Historically this has been done because of worries about statsmodels development upgrading other binaries, e.g., SciPy. I suspect that the changes over the past years have mostly mitigated these worries.

> And pip often destroyed those virtualenv when we had requirements defined

Yes, my impression is that a lot has changed in the `pip` world since 2018 [when (I'm presuming) those decisions were made](https://github.com/statsmodels/statsmodels/pull/4871#issuecomment-414721019). In November of 2020, `pip` launched with [a new dependency resolver](https://pip.pypa.io/en/stable/topics/dependency-resolution/) that no longer installs mutually inconsistent packages. So my understanding is that if you pin your numpy and scipy versions prior to installing statsmodels, then `pip` should respect those pins now. If it can find a version of statsmodels that is consistent with those pins, it will install it. Otherwise, it will error with an informative error message. In other words, it works similarly to conda now.

**Comment by aryarm** [comment]
Having said all of that, I think this discussion about `pip`'s new behavior is actually not relevant to the issue at hand. The statsmodels `setup.py` already defines its version constraints in the `install_requires` [here](https://github.com/statsmodels/statsmodels/blob/e83355ec7d628e3ff8f0efeb44fd6f3ed599c507/setup.py#L404). I wasn't suggesting that those version constraints be changed. Rather, I am just requesting that you also list them in PyPI's JSON API. As I describe above, many tools are starting to depend on that information being available there, so that they don't need to spend time downloading your wheels and extracting the dependency list from the wheels (or worse, the sdist).

**Comment by aryarm** [comment]
> Why should the upload order between sdist and wheel matter?

I'm not sure. And I'm not 100% confident that the upload order is the true cause of the issue, either.

I figured it would be easy to try changing the upload order. But if you decide that it's too difficult given the current setup of your build process, I'm ok with that and can completely understand! And if you decide that you would rather bring this up as an issue with warehouse or twine, I'm ok with that, too.

I'm just trying my best to report that this is an issue, if for no other reason than to warn anybody else who runs into the same problem.

**Comment by josef-pkt** [comment]
just to the history

I ran into the problem more like around 2013  (I think that's around the time pip started to replace easyinstall).
https://stackoverflow.com/questions/15280896/how-to-prevent-tox-from-deleting-installed-packages
oldest issue comment I find
https://github.com/statsmodels/statsmodels/issues/1267#issuecomment-31142004

At that time I had 3 to 5 local virtualenv to do all the testing and debugging of different versions of dependencies.
numpy and especially scipy where still much more buggy and less stable at the time.

All this has changed a lot in the last 9 years

**Comment by aryarm** [comment]
> Why should the upload order between sdist and wheel matter?

[This thread](https://github.com/pypa/warehouse/issues/474#issuecomment-370327255) might also be relevant. In it, they say 
> Metadata extraction currently only happens for the first uploaded artifact, and unlike wheel archives, sdists aren't required to contain metadata in a format that an index server knows how to read.

Later, it's clarified in https://github.com/pypa/warehouse/issues/474#issuecomment-370973132 that the upload order shouldn't matter and that the issue may be due to the use of an older version of twine. However, that thread was written prior to https://github.com/pypa/twine/issues/761, which makes me think the issue probably persists, after all.

**Comment by aryarm** [comment]
This issue seems to persist, even after the 0.13.3 release 😞 
```
> curl -s https://pypi.org/pypi/statsmodels/json | jq ".info.requires_dist"
null
```

**Comment by bashtage** [comment]
I dont know how to fix this, at least not within the current dist. I was careful to upload the tar first.

**Comment by aryarm** [comment]
https://github.com/pypa/twine/issues/761 seems to imply the fix is to upload the wheels before the sdists

Thanks for all that you guys do, and for trying to fix this issue! I really appreciate it.

**Comment by bashtage** [comment]
Looks to be fixed by releasing 0.13.5.



```
 curl -s https://pypi.org/pypi/statsmodels/json | jq ".info.requires_dist"
[
  "pandas (>=0.25)",
  "patsy (>=0.5.2)",
  "packaging (>=21.3)",
  "scipy (>=1.3) ; (python_version > \"3.9\" or platform_system != \"Windows\" or platform_machine != \"x86\") and python_version < \"3.12\"",
  "numpy (>=1.17) ; python_version != \"3.10\" or platform_system != \"Windows\" or platform_python_implementation == \"PyPy\"",
  "numpy (>=1.22.3) ; python_version == \"3.10\" and platform_system == \"Windows\" and platform_python_implementation != \"PyPy\"",
  "scipy (<1.8,>=1.3) ; python_version == \"3.7\"",
  "scipy (<1.9,>=1.3) ; python_version == \"3.8\" and platform_system == \"Windows\" and platform_machine == \"x86\"",
  "scipy (<1.9,>=1.3) ; python_version == \"3.9\" and platform_system == \"Windows\" and platform_machine == \"x86\"",
  "cython (>=0.29.32) ; extra == 'build'",
  "cython (>=0.29.32) ; extra == 'develop'",
  "cython (<3.0.0,>=0.29.32) ; extra == 'develop'",
  "setuptools-scm[toml] (~=7.0.0) ; extra == 'develop'",
  "oldest-supported-numpy (>=2022.4.18) ; extra == 'develop'",
  "matplotlib (>=3) ; extra == 'develop'",
  "colorama ; extra == 'develop'",
  "joblib ; extra == 'develop'",
  "Jinja2 ; extra == 'develop'",
  "pytest (~=7.0.1) ; extra == 'develop'",
  "pytest-randomly ; extra == 'develop'",
  "pytest-xdist ; extra == 'develop'",
  "flake8 ; extra == 'develop'",
  "isort ; extra == 'develop'",
  "pywinpty ; (os_name == \"nt\") and extra == 'develop'",
  "sphinx ; extra == 'docs'",
  "nbconvert ; extra == 'docs'",
  "jupyter-client ; extra == 'docs'",
  "ipykernel ; extra == 'docs'",
  "matplotlib ; extra == 'docs'",
  "nbformat ; extra == 'docs'",
  "numpydoc ; extra == 'docs'",
  "pandas-datareader ; extra == 'docs'"
]
```

**Comment by aryarm** [comment]
hoorah! thank you so much, @bashtage! This will make it much easier for me and other downstream package devs 🙏
